### PR TITLE
Update TLS configuration options

### PR DESCRIPTION
**Pull request by urso**
rename tls to SSL + adapt some fields:

- `insecure` has become `verification_mode`. `insecure: false` is replaced by mode `full` and `insecure: true` by mode `none`.
- renamed `certificate_key` to `key`
- introduced `key_passphrase`
- replaced `min_version` and `max_version` with `supported_protocols` using an array of ok versions. New versions strings: `SSLv3`, `SSLV3.0`, `TLSv1` (use TLSv1.0), `TLSv1.0`, `TLSv1.1` and `TLSv1.2`

Changes:
- rewrite TLS dialer and TLS configuration
- add support for encrypted key files
- use TLS dialer with elasticsearch output (some more manual tests with and without system certificates required)
- update unit/integration tests to use new settings
- update default configuration files to use new ssl settings
- update documentation (docs URL did also change from `tls` to `ssl`)
- update migration script

### CODE REVIEW

File: filebeat/docs/reference/configuration.asciidoc [hunk 1]
```diff
@@ -19,7 +19,7 @@ configuration settings, you need to restart {beatname_uc} to pick up the changes
 * <<redis-output>>
 * <<file-output>>
 * <<console-output>>
-* <<configuration-output-tls>>
+* <<configuration-output-ssl>>
```
Comment (dedemorton) [inline filebeat/docs/reference/configuration.asciidoc]: Just a note that when we change IDs for topics, we need to ask the web team to set up redirects, or our links from forum topics will be broken. That's a good reason to avoid renaming IDs unless (which is probably true in this case) you definitely want the name to change, and it's worth the extra work to get the redirect set up.

File: libbeat/docs/outputconfig.asciidoc [hunk 2]
```diff
@@ -909,18 +909,26 @@ output.elasticsearch:
...
 
-==== TLS Options
+==== SSL Options
 
-You can specify the following options in the `tls` section of the +{beatname_lc}.yml+ config file:
+You can specify the following options in the `ssl` section of the +{beatname_lc}.yml+ config file:
+
+===== enabled
+
+The `enabled` settings can be used to disable the ssl configuration by setting
```
Comment (dedemorton) [inline libbeat/docs/outputconfig.asciidoc]: typo: setting

File: libbeat/docs/outputconfig.asciidoc [hunk 3]
```diff
@@ -930,47 +938,67 @@ The list of root certificates for server verifications. If `certificate_authorit
...
 
-The minimum SSL/TLS version allowed for the encrypted connections. The value must be one of the following:
-`SSL-3.0` for SSL 3, `1.0` for TLS 1.0, `1.1` for TLS 1.1 and `1.2` for TLS 1.2.
+The passphrase used to decrypt an encrypted key stored in the configured `key` file.
 
-The default value is `1.0`.
+===== versions
 
-===== max_version
+List of allowed SSL/TLS versions. If SSL/TLS server decides for protocol verions
```
Comment (dedemorton) [inline libbeat/docs/outputconfig.asciidoc]: Details here are hard to follow. Are you saying that users set this value based on the protocol versions that the SSL/TLS server allows?

File: libbeat/docs/outputconfig.asciidoc [hunk 4]
```diff
@@ -930,47 +938,67 @@ The list of root certificates for server verifications. If `certificate_authorit
...
+not allowed, the connection will be dropped during or after the handshake. The
+setting is a list setting the allowed protocol versions:
```
Comment (dedemorton) [inline libbeat/docs/outputconfig.asciidoc]: Use either "is a list setting that specifies the allowed protocol versions" or "is a list of allowed protocol versions".

File: libbeat/docs/outputconfig.asciidoc [hunk 5]
```diff
@@ -930,47 +938,67 @@ The list of root certificates for server verifications. If `certificate_authorit
...
-`SSL-3.0` for SSL 3, `1.0` for TLS 1.0, `1.1` for TLS 1.1 and `1.2` for TLS 1.2.
+The default value is `[TLSv1.0, TLSv1.1, TLSv1.2]`.
 
-The default value is `1.2`.
+===== verification_mode
 
-===== insecure
+This option controls whether the client verifies server certificates and host
+names. The values `none` and `full` can be used. If `verification_mode` is set
+to `none` , all server host names and certificates are accepted. In this mode,
```
Comment (dedemorton) [inline libbeat/docs/outputconfig.asciidoc]: remove extra space

File: libbeat/docs/outputconfig.asciidoc (hunk 3 above)
Comment (urso) [inline libbeat/docs/outputconfig.asciidoc]: During handshake the client 'advertises' a version and the server might respond with any version it supports. If server responds with version not configured here, we drop the connection. This can happen during handshake (during handshake libs checks version by server in range between min and max configured version) or in some cases after the handshake (e.g. TLSv1.0, and v1.2 configured, but server responds with 1.1).

File: filebeat/docs/migration.asciidoc [hunk 6]
```diff
@@ -442,10 +442,10 @@ Behind the scenes, Filebeat uses a sightly improved protocol for communicating
 with Logstash.
 
 [float]
-=== TLS Is Off by Default
+=== SSL Is Off by Default
 
-If you follow the section on migrating the configuration, you will have TLS
-enabled. However, you must be aware that if the `tls` section is missing from the
+If you follow the section on migrating the configuration, you will have WSSL
```
Comment (ruflin) [inline filebeat/docs/migration.asciidoc]: WSSL?

File: filebeat/filebeat.full.yml [hunk 7]
```diff
@@ -384,34 +384,37 @@ output.elasticsearch:
   # Path to the Elasticsearch 2.x version of the template file.
   #template.versions.2x.path: "${path.config}/filebeat.template-es2x.json"
 
+  # Use SSL settings for HTTPS. Default is true.
+  #ssl.enabled: true
 
-  # TLS configuration. By default is off.
+  # Configure SSL verification mode. If `none` is configured, all server host
```
Comment (ruflin) [inline filebeat/filebeat.full.yml]: s/host/hosts

**Comment by ruflin** [comment]
This needs an update to the CHANGELOG.md file under breaking changes.

File: filebeat/filebeat.full.yml [hunk 8]
```diff
@@ -669,27 +692,38 @@ output.elasticsearch:
   # occurs on the proxy server.
   #proxy_use_local_resolver: false
 
-  # Optional TLS configuration options. TLS is off by default.
+  # Enable SSL support. SSL is automatically enabled, if any SSL setting is set.
+  #ssl.enabled: true
+
+  # Configure SSL verification mode. If `none` is configured, all server host
```
Comment (ruflin) [inline filebeat/filebeat.full.yml]: s/host/hosts/

File: libbeat/docs/outputconfig.asciidoc [hunk 9]
```diff
@@ -248,7 +248,7 @@ After the specified number of retries, the events are typically dropped.
 Some Beats, such as Filebeat, ignore the `max_retries` setting and retry until all
 events are published.
 
-Set `max_retries` to a value less than 0 to retry until all events are published. 
+Set `max_retries` to a value less than 0 to retry until all events are published.
```
Comment (ruflin) [inline libbeat/docs/outputconfig.asciidoc]: Instead of less then 0 I would specify -1. Less guesswork for the person configuring.

File: libbeat/docs/outputconfig.asciidoc [hunk 10]
```diff
@@ -498,7 +498,7 @@ After the specified number of retries, the events are typically dropped.
...
```
Comment (ruflin) [inline libbeat/docs/outputconfig.asciidoc]: See above comment

File: libbeat/docs/outputconfig.asciidoc [hunk 11]
```diff
@@ -590,7 +590,7 @@ After the specified number of retries, the events are typically dropped.
...
```
Comment (ruflin) [inline libbeat/docs/outputconfig.asciidoc]: See above comment.

File: libbeat/docs/outputconfig.asciidoc [hunk 12]
```diff
@@ -633,10 +633,10 @@ Note: If set to 0, no ACKs are returned by Kafka. Messages might be lost silentl
 
 The number of seconds to wait for new events between two producer API calls.
 
-===== tls
+===== ssl
 
-Configuration options for TLS parameters like the root CA for Kibana connections. See
-<<configuration-output-tls>> for more information.
+Configuration options for SSL parameters like the root CA for Kibana connections. See
```
Comment (ruflin) [inline libbeat/docs/outputconfig.asciidoc]: Perhaps I miss something here, but shouldn't Kibana be replaced by elasticsearch?

File: libbeat/docs/outputconfig.asciidoc [hunk 13]
```diff
@@ -752,7 +752,7 @@ After the specified number of retries, the events are typically dropped.
 Some Beats, such as Filebeat, ignore the `max_retries` setting and retry until all
 events are published.
 
-Set `max_retries` to a value less than 0 to retry until all events are published. 
+Set `max_retries` to a value less than 0 to retry until all events are published.
```
Comment (ruflin) [inline libbeat/docs/outputconfig.asciidoc]: See comment above

File: libbeat/docs/outputconfig.asciidoc [hunk 14]
```diff
@@ -890,16 +890,16 @@ The default value is true.
 
 The maximum number of events to buffer internally during publishing. The default is 2048.
 
-Specifying a larger batch size may add some latency and buffering during publishing. However, for Console output, this 
+Specifying a larger batch size may add some latency and buffering during publishing. However, for Console output, this
 setting does not affect how events are published.
 
-Setting `bulk_max_size` to values less than or equal to 0 disables buffering in libbeat. 
+Setting `bulk_max_size` to values less than or equal to 0 disables buffering in libbeat.
```
Comment (ruflin) [inline libbeat/docs/outputconfig.asciidoc]: Just state that it should be set to 0 as they are the same anyways.

File: libbeat/outputs/elasticsearch/config.go [hunk 15]
```diff
@@ -15,7 +15,7 @@ type elasticsearchConfig struct {
 	ProxyURL         string             `config:"proxy_url"`
 	LoadBalance      bool               `config:"loadbalance"`
 	CompressionLevel int                `config:"compression_level" validate:"min=0, max=9"`
-	TLS              *outputs.TLSConfig `config:"tls"`
+	TLS              *outputs.TLSConfig `config:"ssl"`
```
Comment (ruflin) [inline libbeat/outputs/elasticsearch/config.go]: Not required in this PR but we should also rename it internally to SSL for consistency.

File: libbeat/outputs/tls.go [hunk 16]
```diff
@@ -53,167 +84,217 @@ func (c *TLSConfig) Validate() error {
...
+		curves = append(curves, tls.CurveID(id))
+	}
+
+	cert, err := loadCertificate(&config.Certificate)
+	logFail(err)
+
+	cas, errs := loadCertificateAuthorities(config.CAs)
+	logFail(errs...)
+
+	// return config if no error occured
```
Comment (ruflin) [inline libbeat/outputs/tls.go]: Comment is kind of misleading, as I expected that in the next line, the config is returned.

File: libbeat/outputs/transport/tls.go [hunk 17]
```diff
@@ -2,44 +2,262 @@ package transport
...
 	"net"
+	"sync"
 	"time"
 )
 
-func TLSDialer(tlscfg *tls.Config, timeout time.Duration, forward Dialer) Dialer {
-	if tlscfg == nil {
-		return forward
+type TLSConfig struct {
+	Versions []TLSVersion
```
Comment (ruflin) [inline libbeat/outputs/transport/tls.go]: What is there a newline between each line?

File: libbeat/outputs/transport/tls.go [hunk 18]
```diff
@@ -2,44 +2,262 @@ package transport
...
+		v := uint16(version)
+		if v < minVersion {
+			minVersion = v
+		}
+		if v > maxVersion {
+			maxVersion = v
 		}
+	}
+
+	insecure := c.Verification != VerifyFull
```
Comment (ruflin) [inline libbeat/outputs/transport/tls.go]: I suggest we log a warning when insecure (verification == none) is set.

File: libbeat/scripts/migrate_beat_config_1_x_to_5_0.py [hunk 19]
```diff
@@ -1,55 +1,210 @@
...
             elif line == "":
-                outlines.append(line)
+                yield line
+            else:
+                yield line
+                state = "out"
+
+
+@collect_lines
+def migrate_tls_settings(content):
```
Comment (ruflin) [inline libbeat/scripts/migrate_beat_config_1_x_to_5_0.py]: nice

File: libbeat/scripts/migrate_beat_config_1_x_to_5_0.py [hunk 20]
```diff
@@ -143,3 +298,171 @@ def test_migrate_shipper():
 #tags: ["service-X", "web-tier"]
 test:
 """
+
+def test_migrate_tls_settings():
```
Comment (ruflin) [inline libbeat/scripts/migrate_beat_config_1_x_to_5_0.py]: should we move that to a separate file that we don't ship?

**Comment by ruflin** [comment]
LGTM. I left a few minor comments. Would be good to get a second review as this one is quite big.

**Comment by ruflin** [comment]
@urso Did you run `make docs` to check if all the docs still build?

File: filebeat/docs/migration.asciidoc (hunk 6 above)
Comment (urso) [inline filebeat/docs/migration.asciidoc]: ups

File: filebeat/filebeat.full.yml (hunk 7 above)
Comment (urso) [inline filebeat/filebeat.full.yml]: fixed.

File: libbeat/docs/outputconfig.asciidoc (hunk 9 above)
Comment (urso) [inline libbeat/docs/outputconfig.asciidoc]: On todo list for generic output docs updates.

File: libbeat/docs/outputconfig.asciidoc (hunk 12 above)
Comment (urso) [inline libbeat/docs/outputconfig.asciidoc]: oh, interesting. And I just renames TLS to SSL.

Comment (urso) [inline libbeat/docs/outputconfig.asciidoc]: fixed.

File: libbeat/outputs/elasticsearch/config.go (hunk 15 above)
Comment (urso) [inline libbeat/outputs/elasticsearch/config.go]: I did consider renaming structs to SSL, but on the other hand, the module and all stdlib types do use TLS.

File: libbeat/outputs/tls.go (hunk 16 above)
Comment (urso) [inline libbeat/outputs/tls.go]: fixed.

File: libbeat/outputs/transport/tls.go (hunk 17 above)
Comment (urso) [inline libbeat/outputs/transport/tls.go]: wanted to added comments after adding all required fields. Thanks for pointing out.

Comment (urso) [inline libbeat/outputs/transport/tls.go]: done

File: libbeat/scripts/migrate_beat_config_1_x_to_5_0.py (hunk 20 above)
Comment (urso) [inline libbeat/scripts/migrate_beat_config_1_x_to_5_0.py]: maybe in another PR. Added tls migration in line with other migrations.

**Comment by urso** [comment]
> Did you run make docs to check if all the docs still build?

Yep, still builds.

Only issue with docs is (as described by @dedemorton):

> Just a note that when we change IDs for topics, we need to ask the web team to set up redirects, or our links from forum topics will be broken. That's a good reason to avoid renaming IDs unless (which is probably true in this case) you definitely want the name to change, and it's worth the extra work to get the redirect set up.

File: libbeat/outputs/transport/tls.go (hunk 18 above)
Comment (urso) [inline libbeat/outputs/transport/tls.go]: adde warning

**Comment by tsg** [comment]
All looking good, thanks @urso for the great work.

Seems like quite a few redirects are required. We should decide one way or the other so we know if we want to merge the PR as is or adjust it.

**Comment by ruflin** [comment]
I suggest to go with the new urls and do redirects, as ssl is probably also what people search for and want to see in the url.

@urso Needs rebase.

**Comment by urso** [comment]
Yeah, not really happy about all the redirect. But agree with @ruflin about using ssl over tls.
